pkg/llmvalidate/llmvalidate.go

```diff
@@ -0,0 +1,307 @@
+
+	//unmarshall content into []LLMAnswer
+	var answers []LLMAnswer
+	err = json.Unmarshal([]byte(content), &answers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal content: %v", err)
+	}
+
+	return answers, nil
+
+}
+
+func getTextContentFromModelContentResponse(modelResponse *genai.GenerateContentResponse) string {
+	if len(modelResponse.Candidates) == 0 {
+		return ""
+	}
+	content := modelResponse.Candidates[0].Content
+	finalContent := ""
+	for _, part := range content.Parts {
+		finalContent += fmt.Sprint(part)
+	}
+	return finalContent
+}
+
+func getPromptContentForCode(codePath string) ([]string, error) {
+	var prompts []string
+
+	err := filepath.Walk(codePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		relFilePath, err := filepath.Rel(codePath, path)
+		if err != nil {
+			return err
+		}
+
+		extension := filepath.Ext(relFilePath)
+		if !isAllowedExtension(extension) {
+			return nil
+		}
+
+		if isIgnoredFile(relFilePath) {
+			return nil
+		}
+
+		prompt := getPromptContentForFile(codePath, relFilePath)
+		if prompt != "" {
+			prompts = append(prompts, prompt)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return prompts, nil
+}
+
+func isAllowedExtension(extension string) bool {
+	for _, ext := range allowExtensions {
+		if ext == extension {
+			return true
+		}
+	}
+	return false
+}
+
+func isIgnoredFile(file string) bool {
+	flags := fnmatch.FNM_PERIOD | fnmatch.FNM_NOESCAPE
+	for _, pattern := range ignoreList {
+		if fnmatch.Match(pattern, file, flags) {
+			return true
+		}
+	}
+	return false
+}
+
+func getPromptContentForFile(codePath, relFile string) string {
+	content, err := readFileContent(path.Join(codePath, relFile))
+	if err != nil {
+		logme.DebugFln("Error reading file %s: %v", relFile, err)
+		// we are ignoring this error because this might be a non-text file
+		return ""
+	}
+
+	if !utf8.ValidString(content) {
+		return ""
+	}
+
+	if isMinifiedJsFile(content) {
+		return ""
+	}
+
+	if len(content) == 0 {
+		return ""
+	}
+
+	promptContent := fmt.Sprintf(`
+----##----
+Source filename: %s
+Source Content:
+%s
+----##----
+`, relFile, content)
+
+	return promptContent
+}
+
+func isMinifiedJsFile(jsCode string) bool {
+	lines := strings.Split(jsCode, "\n")
+	totalLength := 0
+	for _, line := range lines {
+		totalLength += len(line)
+	}
+	averageLineLength := float64(totalLength) / float64(len(lines))
+	return averageLineLength > 100
+}
+
+func readFileContent(filePath string) (string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	var content strings.Builder
+
+	for scanner.Scan() {
+		if !utf8.ValidString(string(scanner.Text())) {
+			return "", fmt.Errorf("invalid UTF-8 in file %s", filePath)
+		}
+		content.WriteString(scanner.Text() + "\n")
```

Comment:
A couple of things here:
- For efficiency, `scanner.Text()` can be put in a variable so the string isn't allocated twice
- You don't need to cast `scanner.Text()`'s return value to a `string`, because it already returns a `string`
- Rather than writing `scanner.Text() + "\n"` to the `strings.Builder` (which allocates a new string passed to `WriteString`), you can write the string first, and then write `\n` with a separate `WriteRune` call

```suggestion
		text := scanner.Text()
		if !utf8.ValidString(text) {
			return "", fmt.Errorf("invalid UTF-8 in file %s", filePath)
		}
		content.WriteString(text)
		content.WriteRune('\n')
```